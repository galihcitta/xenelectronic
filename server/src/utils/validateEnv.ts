import { port, cleanEnv } from 'envalid';

export const validateEnv = () => {
  cleanEnv(process.env, {
    PORT: port()
  });
};

// https://www.npmjs.com/package/envalid
