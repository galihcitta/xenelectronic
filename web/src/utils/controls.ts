import { Controls } from '../pages/Auth/types';

export const signInControls: Controls = {
  email: {
    label: 'Enter email',
    placeholder: '',
    type: 'email',
    id: 'signin-email',
    name: 'email',
    value: '',
    validationRules: {
      required: true,
      isEmail: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  password: {
    label: 'Enter password',
    placeholder: '',
    type: 'password',
    id: 'signin-password',
    name: 'password',
    value: '',
    validationRules: {
      required: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  }
};

export const signUpControls: Controls = {
  username: {
    label: 'Enter username',
    placeholder: '',
    type: 'text',
    id: 'signup-username',
    name: 'username',
    value: '',
    validationRules: {
      required: true,
      minLength: 4,
      maxLength: 20
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  email: {
    label: 'Enter email',
    placeholder: '',
    type: 'email',
    id: 'signup-email',
    name: 'email',
    value: '',
    validationRules: {
      required: true,
      isEmail: true
    },
    valid: false,
    touched: false,
    errorMsg: ''
  },
  password: {
    label: 'Enter password',
    placeholder: '',
    type: 'password',
    id: 'signup-password',
    name: 'password',
    value: '',
    validationRules: {
      required: true,
      isSecure: true,
      minLength: 8,
      maxLength: 20
    },
    valid: false,
    touched: false,
    errorMsg: ''
  }
};
