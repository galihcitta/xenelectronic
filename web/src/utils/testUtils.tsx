import React from 'react';
import { BrowserRouter, Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../modules/store';
import { createMemoryHistory } from 'history';

const history = createMemoryHistory();

export const withRouter = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <BrowserRouter>
    <Component {...props} />
  </BrowserRouter>
);

export const withHistory = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <Router history={history}>
    <Component {...props} />
  </Router>
);

export const withProvider = <P extends object>(
  Component: React.ComponentType<P>
) => (props: P) => (
  <Provider store={store}>
    <BrowserRouter>
      <Component {...props} />
    </BrowserRouter>
  </Provider>
);
