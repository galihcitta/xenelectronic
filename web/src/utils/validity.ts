import { ValidationRules } from '../pages/Auth/types';

export const validate = (
  value: string,
  controlName: string,
  rules: ValidationRules
) => {
  let isValid = true;
  let errorMsg = '';

  if (rules.required) {
    isValid = value.trim() !== '' && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} field is required`;
  }

  if (rules.minLength) {
    isValid = value.length >= rules.minLength && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} must be at least ${rules.minLength} characters`;
  }

  if (rules.maxLength) {
    isValid = value.length <= rules.maxLength && isValid;
    if (!isValid && !errorMsg.length)
      errorMsg = `${controlName} must be no longer than ${rules.maxLength} characters`;
  }

  if (rules.isEmail) {
    const pattern = /[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*@(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?/;
    isValid = pattern.test(value) && isValid;
    if (!isValid && !errorMsg.length) errorMsg = `email is invalid`;
  }

  if (rules.isSecure) {
    // Minimum eight characters, at least one uppercase letter, one lowercase letter, one number and one special character
    const pattern = /^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{8,}$/;
    isValid = pattern.test(value) && isValid;
    if (!isValid && !errorMsg.length) errorMsg = `password too weak`;
  }

  return {
    isValid,
    errorMsg: errorMsg.charAt(0).toUpperCase() + errorMsg.slice(1)
  };
};
