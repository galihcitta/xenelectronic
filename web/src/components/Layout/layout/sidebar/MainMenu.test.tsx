import { render, fireEvent } from '@testing-library/react';
import { _MainMenu, MainMenuProps } from './MainMenu.component';
import { withRouter } from '../../../../utils/testUtils';

const MainMenuWithRouter = withRouter(_MainMenu);

describe('<CollectionsMenu />', () => {
  const mockHandleCloseSidebar = jest.fn();
  const mockOpenInterior = jest.fn();
  const mockOpenExterior = jest.fn();
  const mockCheckSession = jest.fn();

  const props: MainMenuProps = {
    isAuth: false,
    cartItemsCount: 2,
    showSidebar: false,
    handleCloseSidebar: mockHandleCloseSidebar,
    openInterior: mockOpenInterior,
    openExterior: mockOpenExterior,
    checkSession: mockCheckSession
  };

  const setup = (props: MainMenuProps) => {
    const utils = render(MainMenuWithRouter(props));
    return { ...utils };
  };

  it('renders correctly', () => {
    const { asFragment } = setup(props);
    expect(asFragment()).toMatchSnapshot();
  });

  it('calls `handleCloseSidebar` func', () => {
    const { getByText } = setup(props);
    fireEvent.click(getByText('Home'));
    fireEvent.click(getByText('Account'));
    fireEvent.click(getByText('Cart (2)'));

    expect(mockHandleCloseSidebar).toHaveBeenCalledTimes(3);
  });

  it('calls `openInterior` func', () => {
    const { getByText } = setup(props);
    fireEvent.click(getByText('Interior'));

    expect(mockOpenInterior).toHaveBeenCalled();
  });

  it('calls `openExterior` func', () => {
    const { getByText } = setup(props);
    fireEvent.click(getByText('Exterior'));

    expect(mockOpenExterior).toHaveBeenCalled();
  });

  describe('when `isAuth` prop is true', () => {
    beforeEach(() => {
      props.isAuth = true;
    });

    it('contains `Logout` btn', () => {
      const { getByText } = setup(props);
      expect(getByText('Logout')).toBeInTheDocument();
    });

    it('calls `checkSession` action', () => {
      const { getByText } = setup(props);
      fireEvent.click(getByText('Logout'));

      expect(mockCheckSession).toHaveBeenCalled();
    });
  });
});
