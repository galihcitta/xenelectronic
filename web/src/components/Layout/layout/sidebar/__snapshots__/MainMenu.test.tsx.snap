// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CollectionsMenu /> renders correctly 1`] = `
<DocumentFragment>
  <nav
    class="sc-EHOje ixmxSF"
    data-testid="Wrapper"
  >
    <header
      class="sc-bZQynM lpjVvd"
    >
      <p
        class="sc-gzVnrw dNIQQe"
      >
        Interior & Exterior
      </p>
      <button
        aria-label="close"
        class="sc-htoDjs qBCwI"
        title="close menu"
      >
        <i
          class="far fa-times-circle"
        />
      </button>
    </header>
    <ul
      class="sc-dnqmqq eBryZD"
    >
      <li
        class="sc-bdVaJa jqTRNv"
      >
        <a
          class="sc-htpNat hqricr"
          href="/"
        >
          Home
        </a>
      </li>
      <li
        class="sc-bdVaJa jqTRNv"
      >
        <button
          class="sc-bwzfXH crPdYQ"
        >
          Interior
          <i
            class="fas fa-angle-right"
          />
        </button>
      </li>
      <li
        class="sc-bdVaJa jqTRNv"
      >
        <button
          class="sc-bwzfXH crPdYQ"
        >
          Exterior
          <i
            class="fas fa-angle-right"
          />
        </button>
      </li>
      <li
        class="sc-bdVaJa jqTRNv"
      >
        <a
          class="sc-htpNat hqricr"
          href="/account"
        >
          Account
        </a>
      </li>
      <li
        class="sc-bdVaJa jqTRNv"
      >
        <a
          class="sc-htpNat hqricr"
          href="/cart"
        >
          Cart (2)
        </a>
      </li>
    </ul>
  </nav>
</DocumentFragment>
`;
